src/uu/ls/src/url_escaper.rs

```diff
@@ -0,0 +1,147 @@
+
+        pub fn if_byte_1_if_escape_0(self) -> u16 {
+            self.0 >> 15
+        }
+    }
+}
+
+use byte_or_escape::ByteOrEscape;
+
+pub struct UrlEscaper {
+    table: Box<[ByteOrEscape; 256]>
+}
+
+impl UrlEscaper {
+    pub fn new() -> Self {
+        let mut table = Vec::with_capacity(256);
+        static HEX_BYTES: &[u8; 16] = b"0123456789abcdef";
+        for a in HEX_BYTES.iter().copied() {
+            for b in HEX_BYTES.iter().copied() {
+                table.push(ByteOrEscape::escape(a, b));
+            }
+        }
+        assert_eq!(table.len(), 256);
+        for b in b'0'..=b'9' {
+            table[b as usize] = ByteOrEscape::identity(b);
+        }
+        for b in b'A'..=b'Z' {
+            table[b as usize] = ByteOrEscape::identity(b);
+        }
+        for b in b'a'..=b'z' {
+            table[b as usize] = ByteOrEscape::identity(b);
+        }
+        for b in [b'_', b'-', b'.', b'~'] {
+            table[b as usize] = ByteOrEscape::identity(b);
+        }
+        Self {table: table.into_boxed_slice().try_into().map_err(|_| ()).unwrap()}
+    }
+
+    pub fn into_path_escaper(mut self) -> Self {
+        self.table[b'/' as usize] = ByteOrEscape::identity(b'/');
+        if std::path::MAIN_SEPARATOR != '/' {
+            if let Some(main_separator) = self.table.get_mut(std::path::MAIN_SEPARATOR as usize) {
+                *main_separator = ByteOrEscape::replacement(b'/');
+            }
+        }
+        self
+    }
+
+    pub fn escape<'a>(&self, s: &'a [u8]) -> Cow<'a, str> {
+        let table = &*self.table;
+
+        let mut nonescapes = 0usize;
+        let mut nonidentities = false;
+        for b in s.iter().copied() {
+            if std::path::MAIN_SEPARATOR != '/' && (std::path::MAIN_SEPARATOR as usize) < 256 {
+                nonidentities |= b == (std::path::MAIN_SEPARATOR as u8);
+            }
+            // cannot overflow because it will be at most s.len() since we are summing 0 or 1 values
+            nonescapes = nonescapes.wrapping_add(table[b as usize].if_byte_1_if_escape_0() as usize);
+        }
+
+        if nonescapes == s.len() && !nonidentities {
+            // unsafe: this is safe because the string only has ASCII characters (otherwise it would need escapes)
+            Cow::Borrowed(unsafe {std::str::from_utf8_unchecked(s)})
+        } else {
+            static ERROR: &'static str = "escaped string length too large to fit in memory";
+            // Vec only supports capacity up to isize::MAX, so we must use isize to do all checks
+            let len = isize::try_from(s.len()).expect(ERROR);
+            // can't wrap because nonescapes <= s.len
+            let escapes = len.wrapping_sub(nonescapes as isize);
+            // must check for overflow since we are using unsafe code
+            let size = len.checked_add(escapes).expect(ERROR).checked_add(escapes).expect(ERROR);
+
+            let mem = ManuallyDrop::new(Vec::with_capacity(size as usize)).as_mut_ptr();
```

Comment:
This is never dropped as far as I can tell? Also I would like this much better if it was just using a slice, instead of raw pointers. It's not worth the (hypothetical) performance gain.

Reply:
ManuallyDrop must be used, otherwise the Vec will be freed, leaving the pointer dangling. I split the unsafe code in a separate commit.

Reply:
> ManuallyDrop must be used,

Not if you modify the `Vec` instead of casting it to a pointer and doing manual pointer arithmetic arithmetic.

Reply:
Changed it to modify the vector without recreating it (that should have no effect on the generated code).